src/System.Private.CoreLib/shared/System/Text/EncodingTable.cs

```diff
@@ -0,0 +1,175 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text
+{
+    //
+    // Data table for encoding classes.  Used by System.Text.Encoding.
+    // This class contains two hashtables to allow System.Text.Encoding
+    // to retrieve the data item either by codepage value or by webName.
+    //
+    internal static partial class EncodingTable
+    {
+        private static readonly Dictionary<string, int> s_nameToCodePage = new Dictionary<string, int>(StringComparer.Ordinal);
+        private static readonly Dictionary<int, CodePageDataItem> s_codePageToCodePageData = new Dictionary<int, CodePageDataItem>();
+
+        /*=================================GetCodePageFromName==========================
+        **Action: Given a encoding name, return the correct code page number for this encoding.
+        **Returns: The code page for the encoding.
+        **Arguments:
+        **  name    the name of the encoding
+        **Exceptions:
+        **  ArgumentNullException if name is null.
+        **  internalGetCodePageFromName will throw ArgumentException if name is not a valid encoding name.
+        ============================================================================*/
+
+        internal static int GetCodePageFromName(string name)
+        {
+            if (name == null)
+                throw new ArgumentNullException(nameof(name));
+
+            int code;
+            lock (s_nameToCodePage)
+            {
+                if (s_nameToCodePage.TryGetValue(name, out code))
+                    return code;
+            }
+
+            code = InternalGetCodePageFromName(name);
+
+            lock (s_nameToCodePage)
```

Comment:
It existed previously as it was implemented using synchonized hashtable, see https://github.com/dotnet/coreclr/pull/21735/files#diff-c39b683e1750b4aa61323257ab927dadL46

Reply:
Synchronized Hashtables don't lock on gets:
https://github.com/dotnet/coreclr/blob/d26e31a51d3a97ac925240c134d08418911c457d/src/System.Private.CoreLib/shared/System/Collections/Hashtable.cs#L1382-L1385
because Hashtable is safe for any number of readers concurrent with a single writer (which the locking on writes ensures).

Reply:
Synchronized Hashtable does not have lock for reading. This one has lock for reading.

Reply:
I'd say no as this is already slow path where you look up by a string name and even user encoding providers search is involved before this code is hit

Reply:
What would be better, thread static array?

Reply:
I think `Hashtable` that used to be there before would be fine. (It would not be worse than before.)

Reply:
Or we could just not cache it at all and call InternalGetCodePageFromName all the time, that should not be that much slower that dictionary lookup

Reply:
Depends on whether lower-casing allocates a copy the name or not...